File: Merge_Sort/MergeSort.py
def merge_sort(data_list):

    def merge(left, right):
        result = []
        while len(left) > 0 and len(right) > 0:
            if left[0] <= right[0]:
                result.append(left[0])
                del left[0]
            else:
                result.append(right[0])
                del right[0]

        while len(left) > 0:
            result.append(left[0])
            del left[0]

        while len(right) > 0:
            result.append(right[0])
            del right[0]

        return result

    if len(data_list) <= 1:
        return data_list

    mid = len(data_list) // 2
    left_list = data_list[:mid]
    right_list = data_list[mid:]

    left_list = merge_sort(left_list)
    right_list = merge_sort(right_list)

    return merge(left_list, right_list)

File: SortAlgorithmPractice/src/MergeSort.java

public class MergeSort {

	public static void main(String[] args) {
		int[] arr = RandomNumberMaker.getNumber();
		
		Util.printArray(arr);
		System.out.println("------------------------------------------------------");
		MergeSort.sort(arr, 0, arr.length - 1);
		Util.printArray(arr);
	}
	
	public static void sort(int[] arr, int left, int right) {
		if (left < right) {
			int mid = (left + right) / 2;
			
			sort(arr, left, mid);
			sort(arr, mid + 1, right);
			mergeTwoArea(arr, left, mid, right);
		}
	}
	
	private static void mergeTwoArea(int[] arr, int left, int mid, int right) {
		int leftIdx = left;
		int rightIdx = mid + 1;
		
		int[] temp = new int[right - left + 1];
		int tempIdx = 0;
		while (leftIdx <= mid && rightIdx <= right) {
			if (arr[leftIdx] > arr[rightIdx]) {
				temp[tempIdx++] = arr[rightIdx++];
			} else {
				temp[tempIdx++] = arr[leftIdx++];
			}
		}
		
		while (leftIdx <= mid)
			temp[tempIdx++] = arr[leftIdx++];
		
		while (rightIdx <= right)
			temp[tempIdx++] = arr[rightIdx++];
		
		tempIdx = 0;
		for (int i = left; i <= right; i++) {
			arr[i] = temp[tempIdx++];
		}
	}
}
